### ISSUE TITLE
Can anyone figure out Domo?

### ISSUE THREAD

**Issue by cornfeedhobo** [googlesamples/apps-script-oauth2#81]
I have been trying a few variations and can not figure out how to properly setup the service for Domo.

https://developer.domo.com/docs/authentication/overview-4

**Comment by oshliaer** [comment]
Did you try connect thru curl? I can't do that.

It returns
```json
{
    "status": 401,
    "statusReason": "Internal Server Error",
    "path": "/v1/users",
    "message": "com.domo.bedrock.service.exception.UnauthorizedException: Full authentication is required to access this resource",
    "toe": ""
}
```

**Comment by cornfeedhobo** [comment]
Oh, yeah, maybe I should have mentioned that first, I can do this just fine with `curl`. Getting a token is easy, as are subsequent queries.

**Comment by oshliaer** [comment]
How can I get subsequent queries? In my attempts
```shell
> source AT=access_token
> curl -v -H "Authorization:'bearer $AT'" "https://api.domo.com/v1/users"
...
401
...
```

I'll see if it's possible to receive a token without an authorization with this library..

**Comment by cornfeedhobo** [comment]
Yeah their documentation isn't great. Here is a working example (you'll need jq installed, but can adapt it):
```
export DOMO_OAUTH_ID=...
export DOMO_OAUTH_SECRET=...
export DOMO_TOKEN="$(curl -u "$DOMO_OAUTH_ID:$DOMO_OAUTH_SECRET" "https://api.domo.com/oauth/token?grant_type=client_credentials&scope=data%20user%20dashboard" | jq -r .access_token)"
curl -H "Authorization: bearer $DOMO_TOKEN" "https://api.domo.com/v1/datasets"
```

**Comment by cornfeedhobo** [comment]
I should probably also mention that I have tried using `.setParam('grant_type', 'client_credentials')` as well as using a custom handler like FitBit, but those didn't work either.

**Comment by oshliaer** [comment]
@cornfeedhobo It's strange that their don't require a redirect uri.

Ok. I have a patch for this. But I didn't testing this with rfc6749's capability.

Will you use it as is?

**Comment by cornfeedhobo** [comment]
Is this something you'd be committing to master? What's the patch?

As for giving it a try, sure. But I have no idea how you use a branch of a library if that's what you are looking to do. Google scripts is pretty new to me. I'm just used to normal development...

**Comment by oshliaer** [comment]
@cornfeedhobo 

This lib does not support Client Credentials Grant yet.

As can you see from [Setup](https://github.com/googlesamples/apps-script-oauth2#setup) you may use `/dist` installation instead linking the library. Do it

1. Unbind library from your project
1. Create a gs file with name `Underscore.gs`. Save the content from [./master/dist/Underscore.gs](https://raw.githubusercontent.com/googlesamples/apps-script-oauth2/master/dist/Underscore.gs) to there.
1. Create a gs file with name `OAuth2.gs`. Save the content from [the path](https://gist.github.com/oshliaer/61759916583a71629c8e9c4e0e7f0f0f) to there.
1. Update your code with [Domo](https://gist.github.com/oshliaer/61759916583a71629c8e9c4e0e7f0f0f#file-domo-gs)

It works fine for me https://script.google.com/d/1_cTGT4xUjacppxhfbF-wXuF0tFFk8fYjM1XXIWqGSBVK6sEDK1luYpmb/edit?usp=sharing

**Comment by cornfeedhobo** [comment]
@oshliaer This looks good to me. Did a small scan through the diff, and a test worked. Thanks!  Any hope this will make it to master and a new release pushed?

**Comment by oshliaer** [comment]
@cornfeedhobo

I opened the new issue https://github.com/googlesamples/apps-script-oauth2/issues/82. I'll add Domo API code as an example to there.

It seems that this topic can be closed.

Cheers!

**Comment by oshliaer** [comment]
@cornfeedhobo there are some examples

### POST request
```js
var url = 'https://api.domo.com/v1/users';
    var response = UrlFetchApp.fetch(url, {
      method: 'post',
      'contentType': 'application/json',
      headers: {
        Authorization: 'Bearer ' + service.getAccessToken()
      },
      payload: JSON.stringify({
        'email': email,
        'role': 'Admin',
        'name': 'Name Name',
      })
    });
    Logger.log(response.getResponseCode());
    var result = JSON.parse(response.getContentText());
    Logger.log(JSON.stringify(result, null, 2));
```

### PUT request
```js
var url = 'https://api.domo.com/v1/users/' + id;
var response = UrlFetchApp.fetch(url, {
  method: 'put',
  'contentType': 'application/json',
  headers: {
    Authorization: 'Bearer ' + service.getAccessToken()
  },
  payload: JSON.stringify({
    'id': id,
    'role': 'Admin',
    'email': email,
    'name': 'Anonymous',
  })
});
var result = JSON.parse(response.getContentText());
Logger.log(JSON.stringify(result, null, 2));
```

**Comment by cornfeedhobo** [comment]
Awesome. I started chipping away last night to put together what I need. Thanks again.

**Comment by shashankgaur** [comment]
Hi @oshliaer 

I tried using the [Domo API](https://script.google.com/d/1_cTGT4xUjacppxhfbF-wXuF0tFFk8fYjM1XXIWqGSBVK6sEDK1luYpmb/edit?usp=sharing) code for my own API to request token via client_credentials. I did following steps:

1. Removed OAuth2 library from Resources
2. Copied both OAuth2.gs and Underscore.gs files in the project.
3. Added the getService() function as defined in the Code.gs of Domo API.

However, I still get error that getAccessToken is not defined in the function getService? Am I missing something? I am able to use the normal OAuth2 library but can not get a new token using it.

Thank you
